File: Weekly295/Q3_Steps to Make Array Non-decreasing.py
class Solution:

    def totalSteps(nums)-> int:
        
        n = len(nums)
        dp = [0]*n
        stack = []
        for i in range(n-1,-1,-1):
            while stack and nums[i]>nums[stack[-1]]:
                dp[i]=max(dp[i]+1,dp[stack.pop()])
            stack.append(i)
        return max(dp)

if __name__ == "__main__":
    
    #test1 = Solution([5,3,4,4,7,3,6,11,8,5,11])
    test1 = [5,3,4,4,7,3,6,11,8,5,11]
    test2 = [10,1,2,9,1,2,3,4]

    
    res = Solution.totalSteps(test2)
    print(res)

File: Weekly295/Q3_Steps to Make Array Non-decreasing.cpp
#include <vector>

class Solution{
public:
    int totalSteps(vector<int>& nums){
        int n = nums.size(), res= 0;
        vector<int> dp(n), stack;
        for(int i = n-1;i>=0;--i){
            while(!stack.empty() && nums[i]>nums[sack.back()]){
                dp[i] =max( ++dp[i], dp[stack.back()] );
                stack.pop_back();
                res = max(res,dp[i];)
            }
            stack.push_back();
        }
        return res;
    }
    vector<int> vect{5,3,4,4,7,3,6,11,8,5,11};
    int ans = totalSteps(vect);
    cout << ans;
}